### DISCUSSION TITLE
Why does this `inspector` usage leak memory?

### DISCUSSION THREAD

**Discussion by timfish** [nodejs/help#4819] [Q&A]
I have the following code running in node v20.10.0

`index.mjs`:
```js
import { Session, open } from "node:inspector/promises";

open();

const session = new Session();
session.connect();

let isPaused = false;

session.on("Debugger.resumed", () => {
  isPaused = false;
});

session.on("Debugger.paused", async (event) => {
  isPaused = true;

  for (const frame of event.params.callFrames) {
    const localScope = frame.scopeChain.find((scope) => scope.type === "local");

    if (localScope?.object?.objectId) {
      const props = await session.post("Runtime.getProperties", {
        objectId: localScope.object.objectId,
        ownProperties: true,
      });
    }
  }

  if (isPaused) {
    await session.post("Debugger.resume");
  }
});

await session.post("Debugger.enable");
await session.post("Debugger.setPauseOnExceptions", { state: "all" });

setInterval(() => {
  console.log(`${Math.floor(process.memoryUsage().rss / 1024 / 1024)} MB`);
}, 1000);

setInterval(() => {
  try {
    const a = "nothing";
    throw new Error("Hello");
  } catch {
    //
  }
}, 1);
```

When I run this with `node index.mjs` I get:
```
Debugger listening on ws://127.0.0.1:9229/0f78f4ad-25b9-4947-8713-ec5d92b94ef8
For help, see: https://nodejs.org/en/docs/inspector
60 MB
80 MB
96 MB
107 MB
116 MB
125 MB
140 MB
149 MB
158 MB
166 MB
174 MB
183 MB
193 MB
208 MB
219 MB
228 MB
237 MB
226 MB
239 MB
255 MB
273 MB
282 MB
294 MB
302 MB
312 MB
319 MB
341 MB
350 MB
359 MB
368 MB
377 MB
384 MB
391 MB
```

`isPaused` seems to be required here because by the time we get to end of the `Debugger.paused` callback, the debugger is no longer paused and `await session.post("Debugger.resume")` throws an error saying that it can only be called when the debugger is paused...

My best guess is that this is caused by the fact that I'm passing an async callback for `Debugger.paused` and for whatever reason the debugger continues?

**Comment by timfish** [comment]
It appears that awaiting a promise in the `Debugger.paused` callback causes the debugger to resume and then the context in somehow leaked.

**Comment by marco-ippolito** [comment]
for reference: https://github.com/nodejs/node/issues/51397

**Comment by MohitKawle** [comment]
The await session.post("Debugger.resume") is likely executing asynchronously, and by the time it reaches that line, the debugger may have already resumed. try with   synchronous function and check
